src/main/scala/com/ubirch/webui/models/graph/DefaultGraphClient.scala

```diff
@@ -40,46 +43,78 @@ class DefaultGraphClient(gc: GremlinConnector) extends GraphClient with LazyLogg
     futureCount.map(r => UppState(hwDeviceId, from, to, r.head.toInt))
   }
 
-  override def getLastHashes(hwDeviceId: String, n: Int): Future[List[LastHash]] = {
+  def getLastHashes(hwDeviceId: String, n: Int): Future[List[LastHash]] = {
 
     logger.debug(s"LastHash: Looking for last hashes of $hwDeviceId")
 
-    val futureLastHash = gc.g.V().has(Key[String]("device_id"), hwDeviceId)
-      .value(Key[String]("last_hash")).promise()
+    val futureMaybeLastHash = gc.g
+      .V()
+      .has(Key[String]("device_id"), hwDeviceId)
+      .value(Key[String]("last_hash"))
+      .promise()
+
+    futureMaybeLastHash flatMap { maybeLastHash =>
+      logger.debug(s"LastHash: For $hwDeviceId found one last hash: $maybeLastHash")
+      if (maybeLastHash.isEmpty) {
+        Future.successful(Nil)
+      } else {
+        val futureMaybeLastNUppsValuesMap = lastNHashesTraversal(maybeLastHash.head, n).promise()
+        futureMaybeLastNUppsValuesMap.map { maybeLastNUppsValuesMap =>
+          /* if none are found, then it means that there's less than n upps produced by the device
+          then we check how many were created (should be an inexpensive measure as a number of vertices < n has been created)
+          and we use this number to return this amount of upps queried */
+          if (maybeLastNUppsValuesMap.isEmpty) {
+            val maybeNumberOfUpps = gc.g.
+              V().
+              has(Key[String]("device_id"), hwDeviceId).
+              inE("UPP->DEVICE").
+              count().
+              promise()
+            maybeNumberOfUpps.flatMap { numberOfUpps =>
+              println(s"actualNumber: ${numberOfUpps.head}")
```

Comment:
should the println still be in here? better a logging, with some more info, e.g. also deviceId or remove, not?